Rewrite popup menu positioning logic and fix bugs

girishji opened the pull request:
### Refactor

* Simplified and rewrote horizontal positioning logic (was overly complex).
* Split horizontal and vertical positioning into separate functions.

### Bug fixes

* Fixed missing truncation marker (e.g. `>`) when items were truncated and `pummaxwidth` was not set.
* Fixed occasional extra space being added to menu items.

girishji committed:
Fix & Refactor: Popupmenu positioning

- Simplified and rewrote horizontal positioning logic (was overly
  complex).
- Split horizontal and vertical positioning into separate functions.
- Fixed missing truncation marker (e.g. >) when items were truncated and
  pummaxwidth was not set.
- Fixed occasional extra space being added to menu items.

zeertzjq commented on src/popupmenu.c:
```diff
@@ -80,6 +80,165 @@ pum_compute_size(void)
+}
+
+/*
+ * Calculate horizontal placement for popup menu for insert-mode completion.
```
This is called for cmdline completion as well.

girishji committed:
Improve comment and fix test failure

M  src/popupmenu.c
M  src/testdir/test_popupwin.vim

girishji committed:
Undo test change

M  src/testdir/test_popupwin.vim

girishji committed:
Fix failing test

`Test_popupwin_infopopup_6.dump` and `Test_popupwin_infopopup_7.dump` are
already incorrect: they show the info window but not the popup menu.

This issue occurs only in the test; I cannot reproduce it otherwise.
In `Test_popupmenu_info_border()`, setting of `pumheight=3` causes the
problem. Without `pumheight`, the popup menu displays correctly, but
`pumheight` is needed for the test.

Work around is to add `:set completepopup=border:off` to failing tests.

girishji committed:
Merge branch 'vim:master' into refactor
